File: DeepSpaceJava/src/deepspace/SpecificDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;
import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author samuel
 */
public class SpecificDamage extends Damage {
    SpecificDamage(ArrayList<WeaponType> wl, int s){
        super(wl,s);
    }
    
    @Override
    public Damage copy(){
        return new SpecificDamage(this.getWeapons(),this.getNShields()); 
    }
    
    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
            ArrayList<WeaponType> auxweapons= new ArrayList<WeaponType>();
        auxweapons.addAll(weapons);
        ArrayList<WeaponType> h= new ArrayList<WeaponType>();
        for(int i=0; i<w.size(); ++i){
            if(arrayContainsType(auxweapons, w.get(i).getType())!=-1){
                h.add(auxweapons.get(arrayContainsType(auxweapons, w.get(i).getType())));
                auxweapons.remove(arrayContainsType(auxweapons, w.get(i).getType()));
            }
        }
        Damage d = new SpecificDamage(h, java.lang.Math.min(nShields,s.size()));
        return d;
    }
        
    private int arrayContainsType(ArrayList<WeaponType> w, WeaponType t){
        return w.indexOf(t);
    }
    
    @Override
    void discardWeapon(Weapon w){
        weapons.remove(w.getType());
    }
    
    public ArrayList<WeaponType> getWeapons() {
        return weapons;
    }
    
    @Override
    public boolean hasNoEffect(){
        return(nShields==0 && weapons.size()==0);
    }
    
    SpecificDamageToUI getUIversion(){
        return new SpecificDamageToUI(this);
    }
    
    
}

File: DeepSpaceRuby/lib/SpecificDamage.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Deepspace
  require_relative 'SpecificDamageToUI'
  require_relative'WeaponType'
  class SpecificDamage < Damage
    
    attr_reader:weapons
    
    def initialize(weapons,nshields)
      super(nshields)
      @weapons = Array.new
      for i in (0...weapons.size)
        @weapons.push(weapons[i])
      end
      
    end
    
    def adjust(w,s)
      auxweapons=[]
      for i  in 0...@weapons.size
        auxweapons.push(@weapons[i])
      end
      h=Array.new
      #Sacamos el vector intersección
      for j in 0...w.size
        if arrayContainsType(auxweapons,w[j].type)!=-1
          h.push(auxweapons[arrayContainsType(auxweapons,w[j].type)])
          auxweapons.delete_at(arrayContainsType(auxweapons,w[j].type))
        end
      end
      SpecificDamage.new(h,[@nShields, s.size].min)
    end
    
    def copy()
    SpecificDamage.new(@weapons,@nshields)
  end
  
    def discardWeapon(w)
          @weapons.delete(w.type) 
    end
    
    def getUIversion
      SpecificDamageToUI.new(self)
    end
    
    def hasNoEffect
      return (@nShields==0 && @weapons.size==0)
    end
  
    private
    
    def arrayContainsType(w,t)
      if w.index(t)== nil
        return -1
      else 
        return w.index(t)
      end
    end
    
  end
end
